### DISCUSSION TITLE
Transaction rollback not happening on exception in data fetcher

### DISCUSSION THREAD

**Discussion by hexwit** [graphql-java/graphql-java#2506] [Q&A]
Hi, 

I have following setup: 
Http controller annotated with @Transactional to enable hibernate session, 
and all graphql processing inside this controller method.

At this point everything is fine. Session exists, queries are working. 
But in case if any exception thrown within data fetcher - transaction just committed instead of rollback. 
And data is not consistent of course. So what data session had - everything will be committed. 

I suppose I must manually acquire hibernate session and ask to rollback somewhere in the DataFetcherExceptionHandler. 
But I am not sure 100%.

I have data fetcher exception handler like this: 
```
public class GraphQLExecutionExceptionHandler implements DataFetcherExceptionHandler {
    private Logger log = LoggerFactory.getLogger(getClass());

    @Override
    public DataFetcherExceptionHandlerResult onException(DataFetcherExceptionHandlerParameters handlerParameters) {
        Throwable exception = handlerParameters.getException();
        log.error("Exception during query execution: ", exception);

        GraphQLError graphQLError = GraphQLAPIErrorBuilder.build(exception);

        return DataFetcherExceptionHandlerResult.newResult(graphQLError).build();
    }
}
```

Does graphql engine re-throw exception further or it just swallow it? 
What I am missing here and how to enable transaction rollback? (as by default it is enabled in spring)

Thanks.

**Comment by bbakerman** [comment]
> Does graphql engine re-throw exception further or it just swallow it?

Swallow is the wrong term - but once an exception comes out of DataFetcher, then it hits a `DataFetcherExceptionHandler` how must decide what to do with it.  But default the engine will turn the exception into a graphql error message and continue.

graphql-java cannot execute in a linear fashion like say a PUT on a  REST endpoint because field fetches can and should be async.

```
mutation {
    updateA(...) {
        inner SubSelection
     }
    
    updateB(...) {
       inner subselection
     }
     
}
```

The above does run in a serial fashion HOWEVER each can be async - so `updateA` field fetch completes before `updateB` but they may run on different threads and so on depending on how you set up the code.

So there is no common (they are all done point) easily available.

Adding transactionality to graphql is possible but not in a generic sense which is why we have never tried to offer it.

Sprint offers transaction rollback by default because they do one POST operation at a  time and its all other nothing - eg either that one thing happened or it has an exception.

graphql is not like that by default. (But with work can be massaged into that form)

**Reply by hexwit** [reply]
Thanks for extended reply! 

So, I can assume that working solution would be to put @Transactional to each data fetcher method, to have separate transaction for each call, and add TransactionInterceptor.currentTransactionStatus().setRollbackOnly();  into DataFetcherExceptionHandler to mark current transaction for rollback. 

Is this correct assumption?

**Reply by hexwit** [reply]
Or may be it is possible to enable transactions for each query via instrumentation? I'd like to have some generic approach.
